corehq/util/django_migrations.py

```diff
@@ -110,3 +113,40 @@ def noop_migration():
     Old migrations can be simply removed. `operations = []` is perfectly valid
     """
     return RunPython(RunPython.noop, RunPython.noop)
+
+
+def run_management_command(command_name, *args, required_commit=None, **kwargs):
+    """Return a migration operation that can be used to run a management command from
+    a Django migration. This will give the user directions for running the command
+    manually if the command fails or has since been removed from the codebase"""
+    @skip_on_fresh_install
+    def _run_command(apps, schema_editor):
+        safely_run_management_command(command_name, required_commit, *args, **kwargs)
+
+    return migrations.RunPython(_run_command, reverse_code=migrations.RunPython.noop, elidable=True)
+
+
+def safely_run_management_command(command_name, *args, required_commit=None, custom_message=None, **kwargs):
+    """Helper function to run a management command automatically and abort if there was an error
+    with helpful console output to allow users to run it manually if it fails or
+    has since been removed from the codebase."""
+    try:
+        call_command(command_name, *args, **kwargs)
+    except Exception:
+        traceback.print_exc()
+        print(f"""
+            A migration must be performed before this environment can be upgraded to the latest version
+            of CommCareHQ. This migration is run using the management command {command_name}.
+        """)
+        if required_commit or custom_message:
+            print("")
+            print(custom_message or f"""
+            Run the following commands to run the migration and get up to date:
+
+                commcare-cloud <env> fab setup_limited_release --set code_branch={required_commit}
```

Comment:
Do you think we need to be concerned about the scenario where the limited release does not work because Python requirements are outdated and cannot be installed or something like that?

Reply:
I don't, although I wonder if we're setting up a scenario where these instructions more likely to fail than in the prior workflow? Previously the migration was either retained, meaning it would most likely work with up-to-date requirements, or it was dropped after a suitable period where people had a reasonable chance to upgrade.

I'm wondering if this new workflow will cause us to delete migrations more aggressively, unintentionally putting self-hosters in a more precarious state? The pain will not likely be experienced by people on our team, and our incentives are misaligned with those who are most likely to feel it. I'm not sure what to do about that.

I realize this problem was not entirely absent from the previous workflow since there were scenarios where we instructed people to checkout an older revision

https://github.com/dimagi/commcare-hq/blob/4d90a8b16f3a56e3869070e647c5cc373747b7b0/corehq/form_processor/migrations/0078_blobmeta_migrated_check.py#L15-L18

Reply:
> I'm wondering if this new workflow will cause us to delete migrations more aggressively, unintentionally putting self-hosters in a more precarious state? 

Deleting migrations should never be unintentional. At least this way it make is easier to do while still protecting self-hosters.

I don't think this changes anything in terms of policy or the like - it just standardizes the existing pattern more.

Reply:
Maybe I didn't say that very clearly.

- This new workflow may cause us to delete migrations more aggressively (not unintentionally).
- The unintentional part is that when we delete migrations with this new workflow we may have a false sense of security that the edge cases are handled. We will likely never know if that's true or not, and will not be the most likely ones to experience the pain if not.